liblwgeom/lwgeom_median.c

```diff
@@ -19,156 +19,247 @@
+	 	*/
+		if (hit)
 		{
-			if (distances[i] > 0)
+			double dx = 0;
+			double dy = 0;
+			double dz = 0;
+			double r_inv;
+			for (i = 0; i < npoints; i++)
 			{
-				dx += (points[i].x - curr->x) / distances[i];
-				dy += (points[i].y - curr->y) / distances[i];
-				dz += (points[i].y - curr->z) / distances[i];
+				if (distances[i] > DBL_EPSILON)
+				{
+					dx += (points[i].x - curr->x) / distances[i];
+					dy += (points[i].y - curr->y) / distances[i];
+					dz += (points[i].y - curr->z) / distances[i];
+				}
 			}
-		}
-
-		r_inv = 1.0 / sqrt ( dx*dx + dy*dy + dz*dz );
 
-		alt.x = FP_MAX(0, 1.0 - r_inv)*next.x + FP_MIN(1.0, r_inv)*curr->x;
-		alt.y = FP_MAX(0, 1.0 - r_inv)*next.y + FP_MIN(1.0, r_inv)*curr->y;
-		alt.z = FP_MAX(0, 1.0 - r_inv)*next.z + FP_MIN(1.0, r_inv)*curr->z;
+			r_inv = 1.0 / sqrt ( dx*dx + dy*dy + dz*dz );
 
-		next = alt;
-	}
+			next.x = FP_MAX(0, 1.0 - r_inv)*next.x + FP_MIN(1.0, r_inv)*curr->x;
+			next.y = FP_MAX(0, 1.0 - r_inv)*next.y + FP_MIN(1.0, r_inv)*curr->y;
+			next.z = FP_MAX(0, 1.0 - r_inv)*next.z + FP_MIN(1.0, r_inv)*curr->z;
+		}
 
-	delta = distance3d_pt_pt(curr, &next);
+		delta = distance3d_pt_pt((POINT3D*)curr, (POINT3D*)&next);
 
-	curr->x = next.x;
-	curr->y = next.y;
-	curr->z = next.z;
+		curr->x = next.x;
+		curr->y = next.y;
+		curr->z = next.z;
+	}
 
 	return delta;
 }
 
-static POINT3D
-init_guess(const POINT3D* points, uint32_t npoints)
+static POINT4D
+init_guess(const POINT4D* points, uint32_t npoints)
 {
-	POINT3D guess = { 0, 0, 0 };
+	assert(npoints > 0);
+	POINT4D guess = { 0, 0, 0, 0 };
 	uint32_t i;
 	for (i = 0; i < npoints; i++)
 	{
-		guess.x += points[i].x / npoints;
-		guess.y += points[i].y / npoints;
-		guess.z += points[i].z / npoints;
+		guess.x += points[i].x * points[i].m;
+		guess.y += points[i].y * points[i].m;
+		guess.z += points[i].z * points[i].m;
+		guess.m += points[i].m;
 	}
-
+	guess.x /= guess.m;
+	guess.y /= guess.m;
+	guess.z /= guess.m;
 	return guess;
 }
 
-static POINT3D*
-lwmpoint_extract_points_3d(const LWMPOINT* g, uint32_t* ngeoms)
+static POINT4D*
+lwmpoint_extract_points_4d(const LWMPOINT* g, POINT4D* points, uint32_t* npoints, int* input_ok, int* input_empty)
 {
+	assert(npoints != NULL);
+	assert(input_ok != NULL);
+	assert(*input_ok == LW_TRUE);
+	assert(g->ngeoms > 0);
 	uint32_t i;
 	uint32_t n = 0;
-	int is_3d = lwgeom_has_z((LWGEOM*) g);
+	int has_z = lwgeom_has_z((LWGEOM*) g);
+	int has_m = lwgeom_has_m((LWGEOM*) g);
 
-	POINT3D* points = lwalloc(g->ngeoms * sizeof(POINT3D));
 	for (i = 0; i < g->ngeoms; i++)
 	{
+		assert(n < g->ngeoms);
+		assert(n <= i);
 		LWGEOM* subg = lwcollection_getsubgeom((LWCOLLECTION*) g, i);
 		if (!lwgeom_is_empty(subg))
 		{
-			getPoint3dz_p(((LWPOINT*) subg)->point, 0, (POINT3DZ*) &points[n++]);
-			if (!is_3d)
-				points[n-1].z = 0.0; /* in case the getPoint functions return NaN in the future for 2d */
+			*input_empty = LW_FALSE;
+			if (!getPoint4d_p(((LWPOINT*) subg)->point, 0, &points[n]))
```

Comment:
The previous code excluded empty points from the returned array. It looks like this code retains them?

Reply:
no, it excludes them. it sets a flag to distinguisth 'MULTIPOINT(0 0 0 0)' from 'MULTIPOINT EMPTY'.

Reply:
OK, I see. I was mismatching braces